シェルの初期化関数

ToYeah opened the pull request:
fix #43

### やったこと
シェルの初期化を行う関数`minishell_init`を作りました。mainの中で`loop`前に呼び出してます。
`env -i bash`で`bash`を起動した後、`env`を実行すると、`SHLVL,PWD,OLDPWD`の3つが設定されているので、minishellもそれに倣って3つの環境変数の設定を行ってます。

**PWD**
shell起動時にPWDがあった場合引継ぎ。なければ`getcwd`で作成。
起動時に引き継いだ`PWD`が不正な場合、`getcwd`で補正します。
`PWD`が不正かどうかは、`PWD`に入っている値で`stat`を行い、inode番号の比較で判定しています。`PWD`にシンボリックリンクがふくまれていても`getcwd`の値に補正されます。
補正を行った後、`pwd`コマンドの実装に使う予定の`g_pwd`にコピーを行います。
例：
```
bash-3.2$ pwd
/Users/totaisei/42_minishell
bash-3.2$ export PWD=/bin
bash-3.2$ echo $PWD
/bin
bash-3.2$ bash
bash-3.2$ echo $PWD
/Users/totaisei/42_minishell
bash-3.2$ exit
exit
bash-3.2$ ./minishell 
minishell > echo $PWD
/Users/totaisei/42_minishell
minishell > exit
bash-3.2$ echo $PWD
/bin

bash-3.2$ pwd    
/Users/totaisei/42_minishell
bash-3.2$ mkdir test
bash-3.2$ ln -s test link
bash-3.2$ cd link
bash-3.2$ pwd
/Users/totaisei/42_minishell/link
bash-3.2$ export PWD=a
bash-3.2$ bash
bash-3.2$ echo $PWD
/Users/totaisei/42_minishell/test
bash-3.2$ exit
exit
bash-3.2$ ../minishell 
minishell > echo $PWD
/Users/totaisei/42_minishell/test
```

**OLDPWD**
あれば引継ぎ、値の補正はしません。
無ければ作り、`data`部分は`NULL`にしてあります。

**SHLVL**
shellの深度を表す環境変数のようです。
謎仕様が多いです。
あれば引継ぎ、なければ`data`部分が`1`で作ります
引き継ぐ際、値が不正であれば`1`になります。

|引き継ぐ値|結果|
|---|---|
|`0`未満|`1`|
|`0`から`998`|`+1`した値|
|999|空文字|
|`1000`以上|エラーメッセージを出力した後`1`に補正|
|LONG_MAX,LONG_MINを超える値|`1`|
|数字以外の文字列|`1`|

### 気になっていること
env_utils2.cの命名がダサいかも？
環境変数初期化のテストケースの作り方がわからず全体的なテストがあまりできていません。

ToYeah committed:
feat(utils.h,env_utils2.c): get_envを追加

ToYeah committed:
feat(utils.h, main.c, shell_init.c): minishell_initを追加

ToYeah committed:
Merge branch 'master' into 43-dev-shell-init

ToYeah committed:
feat(libft): ft_atoi_overflow_zeroを追加

ToYeah committed:
refactor(env_utils): env_searchでNULLが帰らないように

ToYeah committed:
feat(shell_init): 環境変数の初期化を実装

ToYeah committed:
refactor(shell_init): normように関数切り分け

ToYeah committed:
style(shlvl_init): norm対応、ファイル切り分け

ToYeah committed:
style(utils.h, env_utils.c): norm対応

ToYeah committed:
feat(shell_init, main.c): g_pwdを追加。pwdの初期化を追加

ToYeah committed:
refactor(shell_init): norm対応

ToYeah committed:
refactor(shlvl_init): 変数の初期化を追加

ToYeah committed:
refactor(libft): libftのnorm対応

ToYeah committed:
refactor(exec.h,utils.h, shell_init): includeを修正, exec.hと重複していたtypedefを修正

nafuka11 commented on srcs/utils/env_utils2.c:
```diff
@@ -0,0 +1,31 @@
+	while (now)
+	{
+		if (ft_strncmp(now->name, name, ft_strlen(name) + 1) == 0)
+			return (now);
```
細かいところですが、 `ft_strcmp` の方がスッキリするかもしれません。

ToYeah replied:
修正しました！
bb6a3b4a0ecd0d3d0c4be0e929397f4a509510bc

nafuka11 commented on srcs/utils/shlvl_init.c:
```diff
@@ -0,0 +1,104 @@
+
+	index = 0;
+	while (str[index] == ' ')
+		index++;
```
bashでは `isspace` を使って空白を読み飛ばしているようです。なので、`\t`, `\n`,    `\v`    `\f`    `\r` も含まれそうです。
`isspace` は libft に入れてもいいかもしれません。

bashの例
```bash
# SHLVLの先頭に\vを含む例
bash-3.2$ export SHLVL=`printf "\v3"`
bash-3.2$ echo $SHLVL

3
bash-3.2$ bash
bash-3.2$ echo $SHLVL
4
```
minishellの例
```bash
bash-3.2$ export SHLVL=`printf "\v3"`
bash-3.2$ ./minishell 
minishell > echo $SHLVL
1
```

bashのソースを見てみると、 variables.c 内の `adjust_shell_level()` でSHLVLの処理を行なっています。
SHLVLの値変換には `legal_number()` を呼び出していて、そこで `strtol()`  を使い、 `isspace` でwhitespaceを読み飛ばしています。

ToYeah replied:
対応しました！
8619dd88c44c0002f048c18cf1fc3e8eff7ed39c
58267e5437db91c42447b2180a753129345b3e55

nafuka11 commented on srcs/utils/shell_init.c:
```diff
@@ -0,0 +1,98 @@
+	return (FALSE);
+}
+
+void	old_pwd_init(void)
+{
+	t_env			*old_pwd_env;
+	extern t_env	*g_envs;
+
+	old_pwd_env = get_env("OLDPWD");
+	if (!old_pwd_env)
+	{
+		if (!(old_pwd_env = malloc(sizeof(t_env))) ||
+			!(old_pwd_env->name = ft_strdup("OLDPWD")))
+		{
+			error_exit(NULL);
+		}
+		old_pwd_env->value = NULL;
+		old_pwd_env->next = NULL;
+		add_env(&g_envs, old_pwd_env);
+	}
+}
```
`OLDPWD` は以前のシェルの値を引き継がないようです。

bashの挙動
```bash
bash-3.2$ cd /
bash-3.2$ cd /bin
bash-3.2$ echo $OLDPWD
/
bash-3.2$ bash
bash-3.2$ echo $OLDPWD
 
```
minishellの挙動
```bash
bash-3.2$ echo $OLDPWD
/
bash-3.2$ {minishellへのディレクトリ}/minishell 
minishell > echo $OLDPWD
/
```

ToYeah replied:
仕様勘違いしてました...
無条件にNULL埋めするようにしました。
d104bebaafe7d9c1f63b9e1596db2b40f76b8e29

nafuka11 reviewed (changes requested):
SHLVL, PWD, OLDPWDの奥深い挙動、勉強になります。
3点気になる点があったので、コメントを残しています。

> env_utils2.cの命名がダサいかも？

- get_*系の関数を集めて、 `env_utils_get.c` とするのはいかがでしょうか。

> 環境変数初期化のテストケースの作り方がわからず全体的なテストがあまりできていません。

- これはテストを実行する前に、任意のコマンドを実行できるようにする必要がありますね。別Issueで対応したいと思います！

ToYeah committed:
refactor(env_utils, expander, build_path): env_utils周りのリファクタリング

search_envの関数名をget_env_dataへ変更
get_env系の関数を集めたenv_utils_getを作成,get_envとget_env_dataのstrncmpをstrcmpへ変更

ToYeah committed:
fix(shell_init): old_pwdを引き継がないよう修正

ToYeah committed:
style(env_utils_get): ヘッダーを修正

ToYeah committed:
feat(libft): ft_isspaceを追加

ft_isprint内に追加
libft.h内のプロトタイプ宣言重複を解消

ToYeah committed:
style(ft_isprint): norm対応

ToYeah committed:
fix(shlvl_init): ft_isspaceの呼出しを追加

ToYeah committed:
fix(shell_init): ft_safe_free_charの呼出しを修正

ToYeah committed:
fix(shell_init): old_pwd内でft_safe_freeを無条件で呼び出されるように

nafuka11 reviewed (approved):
修正を確認しました！
対応いただき、ありがとうございます！！
